MdeModulePkg/Core/Dxe/Misc/MemoryProtectionSupport.c

```diff
@@ -0,0 +1,2853 @@
+  type 'MemoryType' under the configured DXE memory protection policy.
+
+  @param MemoryType       Memory type.
+**/
+UINT64
+GetPermissionAttributeForMemoryType (
+  IN EFI_MEMORY_TYPE  MemoryType
+  );
+
+/**
+  Merge continous memory map entries whose have same attributes.
+
+  @param  MemoryMap              A pointer to the buffer in which firmware places
+                                 the current memory map.
+  @param  MemoryMapSize          A pointer to the size, in bytes, of the
+                                 MemoryMap buffer. On input, this is the size of
+                                 the current memory map.  On output,
+                                 it is the size of new memory map after merge.
+  @param  DescriptorSize         Size, in bytes, of an individual EFI_MEMORY_DESCRIPTOR.
+**/
+VOID
+MergeMemoryMap (
+  IN OUT EFI_MEMORY_DESCRIPTOR  *MemoryMap,
+  IN UINTN                      *MemoryMapSize,
+  IN UINTN                      DescriptorSize
+  );
+
+/**
+  Sort memory map entries based upon PhysicalStart, from low to high.
+
+  @param  MemoryMap              A pointer to the buffer in which firmware places
+                                 the current memory map.
+  @param  MemoryMapSize          Size, in bytes, of the MemoryMap buffer.
+  @param  DescriptorSize         Size, in bytes, of an individual EFI_MEMORY_DESCRIPTOR.
+**/
+VOID
+SortMemoryMap (
+  IN OUT EFI_MEMORY_DESCRIPTOR  *MemoryMap,
+  IN UINTN                      MemoryMapSize,
+  IN UINTN                      DescriptorSize
+  );
+
+/**
+  Set UEFI image memory attributes.
+
+  @param[in]  BaseAddress            Specified start address
+  @param[in]  Length                 Specified length
+  @param[in]  Attributes             Specified attributes
+**/
+VOID
+SetUefiImageMemoryAttributes (
+  IN UINT64  BaseAddress,
+  IN UINT64  Length,
+  IN UINT64  Attributes
+  );
+
+/**
+  Enable NULL pointer detection by changing the attributes of page 0. The assumption is that PEI
+  has set page zero to allocated so this operation can be done safely.
+
+  @retval EFI_SUCCESS       Page zero successfully marked as read protected
+  @retval Other             Page zero could not be marked as read protected
+
+**/
+VOID
+EFIAPI
+EnableNullDetection (
+  VOID
+  );
+
+/**
+  Disable NULL pointer detection.
+**/
+VOID
+EFIAPI
+DisableNullDetection (
+  VOID
+  );
+
+/**
+  Disable NULL pointer detection after EndOfDxe. This is a workaround resort in
+  order to skip unfixable NULL pointer access issues detected in OptionROM or
+  boot loaders.
+
+  @param[in]  Event     The Event this notify function registered to.
+  @param[in]  Context   Pointer to the context data registered to the Event.
+**/
+VOID
+EFIAPI
+DisableNullDetectionEventFunction (
+  EFI_EVENT  Event,
+  VOID       *Context
+  );
+
+extern LIST_ENTRY  mGcdMemorySpaceMap;
+
+// ---------------------------------------
+//     LINKED LIST SUPPORT FUNCTIONS

```

Comment:
These file is unwieldy at the moment. It doesn't need to be done now, but it seems like a `MemoryProtections` directory where each of sections below are in a separate C file would be a start to making it more organized and cohesive.

Reply:
100% agreed. As this gets upstreamed to edk2, I hope to make this much more manageable, but it will take some work.